src/core/border_router/routing_manager.cpp

```diff
@@ -603,47 +684,75 @@ void RoutingManager::EvaluateNat64Prefix(void)
 {
     OT_ASSERT(mIsRunning);
 
-    NetworkData::Iterator            iterator = NetworkData::kIteratorInit;
-    NetworkData::ExternalRouteConfig config;
-    Ip6::Prefix                      smallestNat64Prefix;
+    Error                            error;
+    NetworkData::ExternalRouteConfig preferredNat64PrefixConfig;
 
     LogInfo("Evaluating NAT64 prefix");
 
-    smallestNat64Prefix.Clear();
-    while (Get<NetworkData::Leader>().GetNextExternalRoute(iterator, config) == kErrorNone)
+    error = Get<NetworkData::Leader>().GetPreferredNat64Prefix(preferredNat64PrefixConfig);
+
+    // (Re)advertise NAT64 prefix from this BR when
+    // - no NAT64 prefix exits in Network Data yet
+    // - the preferred NAT64 prefix in Network Data was advertised by this BR
+    if (error == kErrorNotFound || preferredNat64PrefixConfig.GetPrefix() == mAdvertisedNat64Prefix)
     {
-        const Ip6::Prefix &prefix = config.GetPrefix();
+        // Withdraw the legacy infrastructure NAT64 prefix if a new one is provided.
```

Comment:
The logic below is quite complex. I think there are rooms to simplify. 

Maybe a better strategy is:
1. Use GetNat64Prefix to get BR's NAT64 prefix and priority
2. Determine if the BR should add NAT64 prefix
    - Yes if the NAT64 prefix in network data was added by BR, or BR's NAT64 prefix has higher priority than the current NAT64 prefix in network data
    - No otherwise
3. If yes, add BR's NAT64 prefix to network data
    - Do nothing if `mAdvertisedNat64Prefix` is same as BR's NAT64 prefix
    - Withdraw `mAdvertisedNat64Prefix` if it's different from BR's NAT64 prefix
    - Add BR's NAT64 prefix
4. If no, remove `mAdvertisedNat64Prefix` from network data

Note that it's generally a bad idea to check `mAdvertisedNat64Prefix` and behave differently accordingly. `mAdvertisedNat64Prefix` should only be to check if the NAT64 prefix we want to add/remove is in local network data or not to avoid duplicated add/remove.